### PR TITLE
Add an implementation to intercept method calls.

### LINKED ISSUE #1355: [Feature request] To be able to intercept module calls for an existing model.
We are interested in enabling a ‘hook’(‘interceptor’) into calls to Flax modules to add functionality for various purposes. For example, hooks would make it possible to intercept module function calls by changing the input or perturbing the output. It can also be used to access or modify the module gradients. The main motivation for this is to avoid forking existing off-the-shelf  models and create code duplication when it could be avoided. It is also not always feasible to make changes to existing models. For example, a user might want to access a certain intermediate output or gradient for a large number of models. 

Main use cases for this functionality are about perturbing/pruning models or explaining models behavior by looking into gradients. Often the perturbation parameters can be learnable, and injected into model particular layers. For example:

- Differential Masking ([ArXiv](https://arxiv.org/abs/2004.14992)) ([GitHub example with a hook](https://github.com/nicola-decao/diffmask/blob/2385532d4859d6b1608892afe0f827c726aeda75/diffmask/utils/getter_setter.py#L76))
- Parameter space noise for exploration ([ArXiv](https://arxiv.org/pdf/1706.01905.pdf))  ([GitHub example with a hook](https://github.com/iffiX/machin/blob/75b271fb0384a986a6012e17764ceaad65550915/machin/frame/noise/param_space_noise.py#L110))
- SparseML ([Documentation](https://docs.neuralmagic.com/sparseml/))  ([GitHub example with a hook](https://github.com/neuralmagic/sparseml/blob/7c72143a6a8501c7271aa66695be9c0a35c3abb5/src/sparseml/pytorch/optim/mask_pruning.py#L652))
- Class Activation Mapping ([ArXiv](https://arxiv.org/abs/1512.04150))  ([GitHub example with a hook](https://github.com/Project-MONAI/MONAI/blob/12f267c98eabdcd566dff11a3daf931201f04da4/monai/visualize/class_activation_maps.py#L77))

  Note that the hook is a generalization of `capture_intermediates`, in the sense that `capture_intermediates` can be implemented as an interceptor as in the following example for the case of `capture_intermediates=True`:

```
def intercept_method(mdl, fun, *args, **kwargs):
  if fun.__name__ == '__call__':
    y = fun(mdl, *args, **kwargs)
    mdl.sow('intermediates', fun.__name__, y)
    return y
  return fun(mdl, *args, **kwargs)
```
It might be convenient for users if `capture_intermediates` stays, but it could be implemented as an intercept method.

Other frameworks:

- [PyTorch allows](https://pytorch.org/docs/stable/generated/torch.nn.Module.html?highlight=register_forward_hook#torch.nn.Module.register_forward_hook) a user to register hooks using `register_forward_hook`  and `register_forward_pre_hook`. It also has a `register_full_backward_hook` that is called during the backward pass. 
- In the JAX ecosystem, Haiku has an `intercept_methods` [function](https://dm-haiku.readthedocs.io/en/latest/api.html?highlight=getter#haiku.experimental.intercept_methods).

### PR DESCRIPTION
**Pull request by bastings**
# What does this PR do?

Fixes #1355

In particular, it adds an `intercept_method` kwarg to `Module.apply` that allows users to specify a method that can intercept calls to module methods.

## Checklist
- [ ] This PR fixes a minor issue (e.g.: typo or small bug) or improves the docs (you can dismiss the other
      checks if that's the case).
- [X] This change is discussed in a Github issue/
      [discussion](https://github.com/google/flax/discussions) (please add a
      link).
- [X] The documentation and docstrings adhere to the
      [documentation guidelines](https://github.com/google/flax/blob/master/docs/README.md#how-to-write-code-documentation).
- [X] This change includes necessary high-coverage tests.
      (No quality testing = no merge!)

### CODE REVIEW

**Comment by avital** [comment]
Nice @bastings -- this looks very promising! I skimmed it but haven't yet read the implementation in full detail.

A few questions:
1. Can `capture_intermediates` be implemented in terms of `intercept_method`?
2. Can you please remove the formatting changes in `module_test.py`? That'll make it easier to review and less likely to have merge conflicts.

**Comment by bastings** [comment]
hi! 

1) Yes! We left that question for you to decide. We currently kept both to not break things, but you could definitely deprecate `capture_intermediates`. We even wrote a test that shows how to do it.
2) Yes :) the autoformatter did this though, so please consider a separate commit that does the same to make it possible to format code :)

**Comment by jheek** [comment]
>> Yes! We left that question for you to decide. We currently kept both to not break things, but you could definitely deprecate capture_intermediates. We even wrote a test that shows how to do it.

We probably want to keep the separate APIs (at least for now). But I think it's better to keep just one implementation so under the hood I think sow should be implemented as a hook.

**Review by jheek — CHANGES_REQUESTED** [review]

File: flax/linen/module.py [hunk 1]
```diff
@@ -931,6 +946,13 @@ def other_fn(instance, ...):
                treated as mutable: ``bool``: all/no collections are mutable.
                ``str``: The name of a single mutable collection. ``list``: A
                list of names of mutable collections.
+      intercept_method: An optional hook that intercepts all function calls and
+        can be used to modify args and kwargs before calling each original 
+        method, and also to modify its output. The provided method needs to
+        take the form `f(module, original_function, *args, **kwargs)` and is
```
[the remark this answers is not in the document]
Comment (bastings) [reply flax/linen/module.py:952]: Thanks, I went with your suggestion, so the interceptor now takes `f(module, original_function, args, kwargs)`.

File: tests/linen/module_test.py [hunk 2]
```diff
@@ -1232,12 +1287,149 @@ def __call__(self, x):
...
+        grads['inter_grads']['MyModel_0']['layer1']['activation'])
+    # Check the dimentionality of the computed grads with respect to inputs
+    # the the 2nd and 3rd Dense layers, it should match the dimentionalities
+    # of the inputs to the Dense layers.
+    self.assertEqual(grads['inter_grads']['MyModel_0']
+                     ['layer2']['activation'].shape, (10,))
+    self.assertEqual(grads['inter_grads']['MyModel_0']
+                     ['output_layer']['activation'].shape, (10,))
+
+  def test_intercept_method_with_multiple_functions(self):
```
[the remark this answers is not in the document]
Comment (zpolina) [reply tests/linen/module_test.py:1329]: I have added the vmap test, wasn't sure if you wanted to wrap the apply method with vmap or the whole module with decorator, went with apply. But let me know if that works for you.
